File: preface/smallest_free_id/improvement_1/min_free.cpp
#include <ctime>
#include <iostream>
#include <vector>
#include "../generate_data.cpp"
using namespace std;

int min_free(vector<int> &lst, int n) {
  vector<int> flags = vector<int>(n + 1, 0);
  for (auto &i : lst) {
    if (i < n) {
      flags[i] = 1;
    }
  }
  for (size_t i = 0; i < flags.size(); i++) {
    if (flags[i] == 0) {
      return i;
    }
  }
  return -1;
}

int main() {
  // int hundred_thousand = 100000;
  int million = 1000000;
  clock_t total = 0;
  size_t loop_times = 10;
  for (size_t i = 0; i < loop_times; i++) {
    auto lst = generate_data(million, loop_times);
    clock_t start_time = clock();
    int v = min_free(lst, lst.size());
    cout << "result: " << v << endl;
    total += clock() - start_time;
  }
  cout << static_cast<double>(total) / loop_times / CLOCKS_PER_SEC << endl;
}

File: preface/smallest_free_id/improvement_1/min_free.py
import time
import random

HUNDRED_THOUSAND = 100_000
MILLION = 1_000_000


def min_free(lst):
    n = len(lst)
    flags = [False for i in range(n + 1)]
    for i in lst:
        if i < n:
            flags[i] = True
    for i, f in enumerate(flags):
        if f is False:
            return i


def main():
    total = 0
    loop_times = 10
    for _ in range(loop_times):
        data = random.sample(range(0, MILLION + loop_times), k=MILLION)
        start_time = time.clock()
        v = min_free(data)
        print(f"result: {v}")
        total += time.clock() - start_time
    print(f"Average CPU time: {total / loop_times}s")


if __name__ == "__main__":
    main()
